### ISSUE TITLE
Terminal options autocompletion

### ISSUE THREAD

**Issue by HybridDog** [minetest/minetest#6450]
If you type ```minetest --ver``` in a terminal and then press tab, there's no autocompletion. Lots of other programs (e.g. pactl) provide autocompletion. The files enabling bash autocompletion are added in ```/usr/share/bash-completion/completions/```.

**Comment by dsohler** [comment]
It’s rather [impossible to parse the output of `minetest --world list`](https://github.com/minetest/minetest/issues/6452) if you use more than `a-z` as world name …

~~~~
$ minetest --world list
Available worlds:
  'really long ' name' '/home/dirk/.minetest/worlds/really long ' name'
  3x3            '/home/dirk/.minetest/worlds/3x3'
  'test welt'    '/home/dirk/.minetest/worlds/test welt'
  logic          '/home/dirk/.minetest/worlds/logic'
  test           '/home/dirk/.minetest/worlds/test'
~~~~

… so for the sake of simplicity, this is what I got so far …

~~~~
$ minetest --
--address     --go          --name        --random      --trace       --world       
--config      --help        --password    --run         --verbose     --worldname   
--console     --info        --port        --server      --version     
--gameid      --logfile     --quiet       --speedtests  --videomodes  
$ minetest --gameid min
minetest  minimal   
$ minetest --gameid min
~~~~

… using this autocomplete script …

~~~~bash
# file: minetest.sh
# minetest parameter-completion


_minetest() {
    local cur prev all gameIDs
    COMREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"


    # Used with `which` to avoid using aliases. Sets `all` to be a string of
    # all parameters provided by `minetest --help` stripped from white-spaces,
    # newlines and their descriptions.
    all=$($(which minetest) --help \
        | $(which sed) s/'.*\(--[a-z]*\).*'/'\1'/g \
        | $(which grep) -- -- \
        | $(which tr) '\n' ' ')


    # Fetches the game IDs
    gameIDs=$($(which minetest) --gameid list | $(which tr) '\n' ' ')


    case "$cur" in
        -*) COMPREPLY=($(compgen -W "$all" -- $cur));;
    esac


    case "$prev" in
        --gameid ) COMPREPLY=($(compgen -W "$gameIDs" -- $cur));;
    esac
}


complete -F _minetest minetest
~~~~

**Comment by paramat** [comment]
Autocompletion is somewhat of a luxury and low priority.

**Comment by dsohler** [comment]
Well, I cobbled together the script in my free time. Unpaid. And I’m not even a MT dev. Many of the completion scripts are for unpaid one-man projects. Not having one because “muh unpaid” isn’t a valid reason.

It took a few minutes to find out that the information returned by Minetest is not properly parsable but It’s a good start I guess.

**Comment by HybridDog** [comment]
> It’s a good start I guess.

I like it, it even doesn't need to be updated when a new parameter is added to minetest.

To get the world names, you could use the fact that worldnames can't contain `/`, so `'/` likely is the beginning of the world path.

**Comment by dsohler** [comment]
> it even doesn't need to be updated when a new parameter is added to minetest.

Yes, the list of parameters is there with `minetest --help` why to re-implement it? :smile: Too bad this is all we can currently get because there is no way to properly parse the world list output.

Btw. I just added a repo to my Minetest Github account.

https://github.com/4w/bash-autocomplete

**Comment by nerzhul** [comment]
if you have a working script we can embed it into misc/ folder for packagers it's a good idea

**Comment by dsohler** [comment]
@nerzul The linked script is working (but lacks some features because Minetest lacks parsable listing of world names and/or world folders).

**Comment by HybridDog** [comment]
The links are broken. Here's a perhaps old version of the script which I downloaded a long time ago: ![minetest sh](https://user-images.githubusercontent.com/3192173/144631733-a0a41490-fc28-40ff-8715-f21d22ba79e8.jpg) (.jpg extension added against github file name complaints)

**Comment by AFCMS** [comment]
Just noticing this issue, so here is a ZSH completion script I wrote a while ago:

https://gist.github.com/AFCMS/adb2035d77ad1a37918a7714946e2dbf

Hope it will be helpful.

**Comment by wsor4035** [comment]
(please ignore this, unsubscribing)

**Comment by AFCMS** [comment]
I am working on a PowerShell completion script.

https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/register-argumentcompleter?view=powershell-7.4

My wish list for completion scripts:
- [ ] Bash (fixed by PR)
- [ ] ZSH (my above script isn't perfect and isn't 100% up to date)
- [ ] PowerShell
